src/file_upload.rs

```diff
@@ -121,37 +116,99 @@ fn handle_multipart(
 /// server root directory. Any path which will go outside of this directory is considered
 /// invalid.
 /// This method returns future.
-pub fn upload_file(req: &HttpRequest<crate::MiniserveConfig>) -> FutureResponse<HttpResponse> {
+pub fn upload_file(
+    req: &HttpRequest<crate::MiniserveConfig>,
+    default_color_scheme: ColorScheme,
+) -> FutureResponse<HttpResponse> {
     let return_path = if let Some(header) = req.headers().get(header::REFERER) {
         header.to_str().unwrap_or("/").to_owned()
     } else {
         "/".to_string()
     };
-    let app_root_dir = if let Ok(dir) = req.state().path.canonicalize() {
-        dir
-    } else {
-        return Box::new(create_error_response("Internal server error", &return_path));
-    };
-    let path = match Query::<QueryParameters>::extract(req) {
+
+    let (path, sort_method, sort_order, color_scheme) = match Query::<QueryParameters>::extract(req)
+    {
         Ok(query) => {
-            if let Ok(stripped_path) = query.path.strip_prefix(Component::RootDir) {
-                stripped_path.to_owned()
+            let sort_param = query.sort;
+            let order_param = query.order;
+            let theme_param = query.theme.unwrap_or(default_color_scheme);
+
+            if let Some(path) = query.path.clone() {
+                if let Ok(stripped_path) = path.strip_prefix(Component::RootDir) {
+                    (
+                        stripped_path.to_owned(),
+                        sort_param,
+                        order_param,
+                        theme_param,
+                    )
+                } else {
+                    (path.clone(), sort_param, order_param, theme_param)
+                }
             } else {
-                query.path.clone()
+                let err = ContextualError::InvalidHTTPRequestError(
+                    "Missing query parameter 'path'".to_string(),
+                );
+                return Box::new(create_error_response(
+                    &err.to_string(),
+                    StatusCode::BAD_REQUEST,
+                    &return_path,
+                    sort_param,
+                    order_param,
+                    theme_param,
+                    default_color_scheme,
+                ));
             }
         }
-        Err(_) => {
+        Err(e) => {
+            let err = ContextualError::InvalidHTTPRequestError(e.to_string());
             return Box::new(create_error_response(
-                "Unspecified parameter path",
+                &err.to_string(),
+                StatusCode::BAD_REQUEST,
                 &return_path,
-            ))
+                None,
+                None,
+                default_color_scheme,
```

Comment:
This too should return the currently active color scheme, I think.

Reply:
This one I won't be able to fix. It's the error occurring when trying to extract the query parameters from the URI. I can't retrieve the active color scheme if the extracting fails :P

Reply:
Ah, hah yeah fair enough.